### PR TITLE
Kernel updates 2024-05-27 (incl 6.10-rc1)

### PR DESCRIPTION
**Pull request by alyssais**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by K900 — APPROVED** [review]
Diff LGTM

**Review by Ma27 — CHANGES_REQUESTED** [review]
The Rust patch seems already applied for `/nix/store/z4sf9jia1rzqlkpgbx3985glaxk8hy9g-linux-6.10-rc1.drv`:

<details><summary>error log</summary>

```
       last 25 log lines:
       > Perhaps you used the wrong -p or --strip option?
       > The text leading up to this was:
       > --------------------------
       > |diff --git a/rust/alloc/vec/mod.rs b/rust/alloc/vec/mod.rs
       > |index 220fb9d6f45b3f..0be27fff4554a1 100644
       > |--- a/rust/alloc/vec/mod.rs
       > |+++ b/rust/alloc/vec/mod.rs
       > --------------------------
       > File to patch:
       > Skip this patch? [y]
       > Skipping patch.
       > 15 out of 15 hunks ignored
       > patching file rust/kernel/lib.rs
       > Reversed (or previously applied) patch detected!  Assume -R? [n]
       > Apply anyway? [n]
       > Skipping patch.
       > 2 out of 2 hunks ignored -- saving rejects to file rust/kernel/lib.rs.rej
       > patching file scripts/Makefile.build
       > Reversed (or previously applied) patch detected!  Assume -R? [n]
       > Apply anyway? [n]
       > Skipping patch.
       > 1 out of 1 hunk ignored -- saving rejects to file scripts/Makefile.build.rej
       > patching file scripts/min-tool-version.sh
       > Hunk #1 FAILED at 33.
       > 1 out of 1 hunk FAILED -- saving rejects to file scripts/min-tool-version.sh.rej
       For full logs, run 'nix log /nix/store/2rarf83b3j9202d533i8bdzdspk5hd1a-linux-config-6.10-rc1.drv'.
```

</details>

**Comment by alyssais** [comment]
Ah right, of course.

**Comment by Ma27** [comment]
fwiw rest of the changed kernels seem good though.

**Review by Nanotwerp — APPROVED** [review]
LGTM! The removal of the now-redundant `Rust` patch and the renaming of the config options fixed my compilation of the kernel.

**Comment by alyssais** [comment]
linux_testing doesn't build for riscv64, I assume because of https://git.kernel.org/linus/e79dfcbfb902a99268cc8022031461da7a8e2bc8.  It will now unconditionally try install Image.zst, which we haven't built because platforms.nix says we want Image on Linux, and that can't just be changed to Image.zst because that doesn't exist in older kernels.

File: pkgs/os-specific/linux/kernel/common-config.nix [hunk 1]
```diff
@@ -927,8 +929,9 @@ let
       # i686 issues: https://github.com/NixOS/nixpkgs/pull/117961#issuecomment-812106375
       useZstd = stdenv.buildPlatform.is64bit && versionAtLeast version "5.9";
     in {
-      KERNEL_XZ            = mkIf (!useZstd) yes;
-      KERNEL_ZSTD          = mkIf useZstd yes;
+      KERNEL_UNCOMPRESSED  = mkIf stdenv.hostPlatform.isRiscV yes;
```
Comment (Ma27) [inline pkgs/os-specific/linux/kernel/common-config.nix:932]: A comment on the rationale for riscv may be useful here.

**Review by Ma27 — APPROVED** [review]
May I ask why we assume that Risc-V uses an uncompressed kernel? Not sure I understand that part.

Approving anyways, since this is not my area of expertise and I don't want to block an otherwise valid PR because I probably don't know enough about that.

**Comment by alyssais** [comment]
> May I ask why we assume that Risc-V uses an uncompressed kernel?

From the commit message:

We assume RISC-V uses an uncompressed kernel in lib/systems/platforms.nix.  Since this is the first Linux version to support either xz or zstd compressed kernels on RISC-V, we can't change platforms.nix to match the other platforms without breaking builds of older kernel versions, so let's just preserve the old behaviour.

> We could in future decide to use gzip-compressed kernels on RISC-V, which is supported going back further.

File: pkgs/os-specific/linux/kernel/common-config.nix (hunk 1 above)
Comment (alyssais) [reply pkgs/os-specific/linux/kernel/common-config.nix:933]: Added.

**Comment by Ma27** [comment]
Sorry, understood. Thanks!

File: pkgs/os-specific/linux/kernel/common-config.nix (hunk 1 above)
Comment (alyssais) [reply pkgs/os-specific/linux/kernel/common-config.nix:933]: Done.

**Comment by alyssais** [comment]
@blitz @Ma27 @yu-re-ka I've just noticed that this has sadly regressed Rust on aarch64.  It now uses the aarch64-unknown-none target instead of the default one.  We could just build support for that into our default rustc, except that the build fails because it can't generate documentation for std for that target (because there is no std).  I think the only good way to resolve this is going to be for Rust's build system to be fixed to skip trying to generate std documentation only for no_std targets, or to at least provide that option.  I don't think I'm going to be able to work on that for at least a while.

**Pull request merged**